File: stockGame-be/src/auth/services/auth.service.ts
import { Injectable, HttpException, HttpStatus } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { User, UserRole } from '../../users/entities/user.entity';
import { UserSession } from '../entities/user-session.entity';
import { JwtService } from './jwt.service';
import { KakaoOAuthService, KakaoUserInfo } from './kakao-oauth.service';
import { GoogleOAuthService, GoogleUserInfo } from './google-oauth.service';

export interface AuthTokens {
  accessToken: string;
  refreshToken: string;
  user: {
    id: string;
    name: string;
    nickname: string;
    email?: string;
    role: string;
    balance: number;
  };
}

@Injectable()
export class AuthService {
  constructor(
    @InjectRepository(User)
    private readonly userRepository: Repository<User>,
    @InjectRepository(UserSession)
    private readonly userSessionRepository: Repository<UserSession>,
    private readonly jwtService: JwtService,
    private readonly kakaoOAuthService: KakaoOAuthService,
    private readonly googleOAuthService: GoogleOAuthService,
  ) {}

  getKakaoAuthUrl(): string {
    const url = `https://kauth.kakao.com/oauth/authorize?client_id=${process.env.KAKAO_CLIENT_ID}&redirect_uri=${process.env.KAKAO_REDIRECT_URI}&response_type=code`;
    return url;
  }

  getGoogleAuthUrl(): string {
    return this.googleOAuthService.getGoogleAuthUrl();
  }

  async kakaoLogin(authorizationCode: string): Promise<AuthTokens> {
    // 1. 카카오 액세스 토큰 획득
    const kakaoAccessToken =
      await this.kakaoOAuthService.getAccessToken(authorizationCode);

    // 2. 카카오 사용자 정보 획득
    const kakaoUserInfo =
      await this.kakaoOAuthService.getUserInfo(kakaoAccessToken);

    // 3. 사용자 찾기 또는 생성
    let user = await this.userRepository.findOne({
      where: { kakaoId: kakaoUserInfo.id.toString() },
    });

    if (!user) {
      // 새 사용자 생성
      user = await this.createUserFromKakao(kakaoUserInfo);
    } else {
      // 마지막 로그인 시간 업데이트
      user.lastLoginAt = new Date();
      await this.userRepository.save(user);
    }

    // 4. 기존 세션 정리 (보안을 위해)
    await this.cleanupOldSessions(user.id);

    // 5. JWT 토큰 생성
    const accessToken = this.jwtService.generateAccessToken({
      sub: user.id,
      email: user.email,
      role: user.role,
    });

    const refreshToken = this.jwtService.generateRefreshToken({
      sub: user.id,
    });

    // 6. 리프레시 토큰 저장
    await this.saveRefreshToken(user.id, refreshToken);

    return {
      accessToken,
      refreshToken,
      user: {
        id: user.id,
        name: user.name,
        nickname: user.nickname,
        email: user.email,
        role: user.role,
        balance: user.balance,
      },
    };
  }

  async googleLogin(authorizationCode: string): Promise<AuthTokens> {
    // 1. 구글 액세스 토큰 획득
    const googleAccessToken =
      await this.googleOAuthService.getAccessToken(authorizationCode);

    // 2. 구글 사용자 정보 획득
    const googleUserInfo =
      await this.googleOAuthService.getUserInfo(googleAccessToken);

    // 3. 사용자 찾기 또는 생성
    let user = await this.userRepository.findOne({
      where: { googleId: googleUserInfo.id },
    });

    if (!user) {
      // 새 사용자 생성
      user = await this.createUserFromGoogle(googleUserInfo);
    } else {
      // 마지막 로그인 시간 업데이트
      user.lastLoginAt = new Date();
      await this.userRepository.save(user);
    }

    // 4. 기존 세션 정리 (보안을 위해)
    await this.cleanupOldSessions(user.id);

    // 5. JWT 토큰 생성
    const accessToken = this.jwtService.generateAccessToken({
      sub: user.id,
      email: user.email,
      role: user.role,
    });

    const refreshToken = this.jwtService.generateRefreshToken({
      sub: user.id,
    });

    // 6. 리프레시 토큰 저장
    await this.saveRefreshToken(user.id, refreshToken);

    return {
      accessToken,
      refreshToken,
      user: {
        id: user.id,
        name: user.name,
        nickname: user.nickname,
        email: user.email,
        role: user.role,
        balance: user.balance,
      },
    };
  }

  private async createUserFromKakao(
    kakaoUserInfo: KakaoUserInfo,
  ): Promise<User> {
    const user = this.userRepository.create({
      email: kakaoUserInfo.kakao_account.email,
      name: kakaoUserInfo.kakao_account.profile.nickname,
      nickname: kakaoUserInfo.kakao_account.profile.nickname,
      profileImage: kakaoUserInfo.kakao_account.profile.profile_image_url,
      kakaoId: kakaoUserInfo.id.toString(),
      role: UserRole.USER,
      balance: 10000000, // 1000만원
      lastLoginAt: new Date(),
    });

    return await this.userRepository.save(user);
  }

  private async createUserFromGoogle(
    googleUserInfo: GoogleUserInfo,
  ): Promise<User> {
    const user = this.userRepository.create({
      email: googleUserInfo.email,
      name: googleUserInfo.name,
      nickname: googleUserInfo.name,
      profileImage: googleUserInfo.picture,
      googleId: googleUserInfo.id,
      role: UserRole.USER,
      balance: 10000000, // 1000만원
      lastLoginAt: new Date(),
    });

    return await this.userRepository.save(user);
  }

  private async cleanupOldSessions(userId: string): Promise<void> {
    await this.userSessionRepository.delete({ userId });
  }

  private async saveRefreshToken(
    userId: string,
    refreshToken: string,
  ): Promise<void> {
    const expiresAt = new Date();
    expiresAt.setDate(expiresAt.getDate() + 7); // 7일 후 만료

    const session = this.userSessionRepository.create({
      userId,
      refreshToken,
      expiresAt,
    });

    await this.userSessionRepository.save(session);
  }

  async refreshToken(refreshToken: string): Promise<AuthTokens> {
    // 1. 리프레시 토큰 검증
    const session = await this.userSessionRepository.findOne({
      where: { refreshToken },
    });

    if (!session || session.expiresAt < new Date()) {
      throw new HttpException(
        '유효하지 않은 리프레시 토큰입니다.',
        HttpStatus.UNAUTHORIZED,
      );
    }

    // 2. 사용자 정보 조회
    const user = await this.userRepository.findOne({
      where: { id: session.userId },
    });

    if (!user) {
      throw new HttpException(
        '사용자를 찾을 수 없습니다.',
        HttpStatus.UNAUTHORIZED,
      );
    }

    // 3. 새로운 토큰 생성
    const newAccessToken = this.jwtService.generateAccessToken({
      sub: user.id,
      email: user.email,
      role: user.role,
    });

    const newRefreshToken = this.jwtService.generateRefreshToken({
      sub: user.id,
    });

    // 4. 기존 세션 삭제 및 새 세션 저장
    await this.userSessionRepository.delete({ id: session.id });
    await this.saveRefreshToken(user.id, newRefreshToken);

    return {
      accessToken: newAccessToken,
      refreshToken: newRefreshToken,
      user: {
        id: user.id,
        name: user.name,
        nickname: user.nickname,
        email: user.email,
        role: user.role,
        balance: user.balance,
      },
    };
  }

  async logout(userId: string): Promise<void> {
    await this.userSessionRepository.delete({ userId });
  }

  async validateUser(userId: string): Promise<User | null> {
    return await this.userRepository.findOne({
      where: { id: userId },
    });
  }
}

File: stockGame-be/src/main/java/kr/knu/getit/stockgame/service/AuthService.java
package kr.knu.getit.stockgame.service;

import kr.knu.getit.stockgame.dto.AuthDto;
import kr.knu.getit.stockgame.entity.User;
import kr.knu.getit.stockgame.entity.UserRole;
import kr.knu.getit.stockgame.entity.UserSession;
import kr.knu.getit.stockgame.repository.UserRepository;
import kr.knu.getit.stockgame.repository.UserSessionRepository;
import kr.knu.getit.stockgame.security.JwtUtil;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.Objects;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final UserSessionRepository sessionRepository;
    private final JwtUtil jwtUtil;
    private final KakaoOAuthService kakaoOAuthService;
    private final GoogleOAuthService googleOAuthService;

    @Value("${app.jwt.refresh-expires-in:7d}")
    private String refreshExpiresIn;

    public String getKakaoAuthUrl() {
        return kakaoOAuthService.getAuthUrl();
    }

    public String getGoogleAuthUrl() {
        return googleOAuthService.getAuthUrl();
    }

    @Transactional
    public AuthDto.AuthTokensResponse kakaoLogin(String code) {
        String accessToken = kakaoOAuthService.getAccessToken(code);
        KakaoOAuthService.KakaoUserInfo info = kakaoOAuthService.getUserInfo(accessToken);

        User user = userRepository.findByKakaoId(info.id())
                .orElseGet(() -> createUserFromKakao(info));

        if (user.getLastLoginAt() == null || user.getLastLoginAt().isBefore(Instant.now().minus(1, ChronoUnit.MINUTES))) {
            user.setLastLoginAt(Instant.now());
            userRepository.save(user);
        }

        sessionRepository.deleteByUserId(user.getId());
        String refreshToken = jwtUtil.generateRefreshToken(user.getId());
        saveRefreshToken(user.getId(), refreshToken);

        String jwtAccess = jwtUtil.generateAccessToken(
                user.getId(),
                user.getEmail(),
                user.getRole().name()
        );

        return toAuthResponse(user, jwtAccess, refreshToken);
    }

    @Transactional
    public AuthDto.AuthTokensResponse googleLogin(String code) {
        String accessToken = googleOAuthService.getAccessToken(code);
        GoogleOAuthService.GoogleUserInfo info = googleOAuthService.getUserInfo(accessToken);

        User user = userRepository.findByGoogleId(info.id())
                .orElseGet(() -> createUserFromGoogle(info));

        if (user.getLastLoginAt() == null || user.getLastLoginAt().isBefore(Instant.now().minus(1, ChronoUnit.MINUTES))) {
            user.setLastLoginAt(Instant.now());
            userRepository.save(user);
        }

        sessionRepository.deleteByUserId(user.getId());
        String refreshToken = jwtUtil.generateRefreshToken(user.getId());
        saveRefreshToken(user.getId(), refreshToken);

        String jwtAccess = jwtUtil.generateAccessToken(
                user.getId(),
                user.getEmail(),
                user.getRole().name()
        );

        return toAuthResponse(user, jwtAccess, refreshToken);
    }

    private User createUserFromKakao(KakaoOAuthService.KakaoUserInfo info) {
        long count = userRepository.count();
        UserRole role = count == 0 ? UserRole.ADMIN : UserRole.USER;
        User user = User.builder()
                .email(info.email())
                .name(info.nickname())
                .nickname(info.nickname())
                .profileImage(info.profileImageUrl())
                .kakaoId(info.id())
                .role(role)
                .balance(new BigDecimal("10000000"))
                .lastLoginAt(Instant.now())
                .isLeaderboardVisible(true)
                .build();
        return Objects.requireNonNull(userRepository.save(user));
    }

    private User createUserFromGoogle(GoogleOAuthService.GoogleUserInfo info) {
        long count = userRepository.count();
        UserRole role = count == 0 ? UserRole.ADMIN : UserRole.USER;
        User user = User.builder()
                .email(info.email())
                .name(info.name())
                .nickname(info.name())
                .profileImage(info.picture())
                .googleId(info.id())
                .role(role)
                .balance(new BigDecimal("10000000"))
                .lastLoginAt(Instant.now())
                .isLeaderboardVisible(true)
                .build();
        return Objects.requireNonNull(userRepository.save(user));
    }

    private void saveRefreshToken(String userId, String refreshToken) {
        long days = 7;
        if (refreshExpiresIn != null && refreshExpiresIn.endsWith("d")) {
            days = Long.parseLong(refreshExpiresIn.replace("d", ""));
        }
        UserSession session = Objects.requireNonNull(UserSession.builder()
                .userId(Objects.requireNonNull(userId))
                .refreshToken(Objects.requireNonNull(refreshToken))
                .expiresAt(Instant.now().plus(days, ChronoUnit.DAYS))
                .build());
        sessionRepository.save(session);
    }

    @Transactional
    public AuthDto.AuthTokensResponse refreshToken(String refreshToken) {
        UserSession session = sessionRepository.findByRefreshToken(refreshToken)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.UNAUTHORIZED, "유효하지 않은 리프레시 토큰입니다."));

        if (session.getExpiresAt().isBefore(Instant.now())) {
            sessionRepository.delete(session);
            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "유효하지 않은 리프레시 토큰입니다.");
        }

        User user = Objects.requireNonNull(userRepository.findById(session.getUserId())
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.UNAUTHORIZED, "사용자를 찾을 수 없습니다.")));

        sessionRepository.delete(session);
        String newRefresh = jwtUtil.generateRefreshToken(user.getId());
        saveRefreshToken(user.getId(), newRefresh);
        String newAccess = jwtUtil.generateAccessToken(user.getId(), Objects.requireNonNull(user.getEmail()), user.getRole().name());

        return toAuthResponse(user, newAccess, newRefresh);
    }

    public Optional<User> validateUser(String userId) {
        return userRepository.findById(userId);
    }

    @Transactional
    public AuthDto.AuthTokensResponse createTokensForUser(User user) {
        String userId = Objects.requireNonNull(user.getId());
        sessionRepository.deleteByUserId(userId);
        String refreshToken = jwtUtil.generateRefreshToken(userId);
        saveRefreshToken(userId, refreshToken);
        String accessToken = jwtUtil.generateAccessToken(
                userId,
                Objects.requireNonNull(user.getEmail()),
                user.getRole().name()
        );
        return toAuthResponse(user, accessToken, refreshToken);
    }

    private static AuthDto.AuthTokensResponse toAuthResponse(User user, String accessToken, String refreshToken) {
        return AuthDto.AuthTokensResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .user(AuthDto.AuthTokensResponse.UserInfo.builder()
                        .id(user.getId())
                        .name(user.getName())
                        .nickname(user.getNickname())
                        .email(user.getEmail())
                        .role(user.getRole().name())
                        .balance(user.getBalance())
                        .build())
                .build();
    }
}
